apps/designer/app/canvas/shared/use-drag-drop.ts

```diff
@@ -0,0 +1,289 @@
+};
+
+export type DragStartPayload = {
+  origin: "panel" | "canvas";
+  dragItem: BaseInstance;
+};
+
+export type DragEndPayload = {
+  origin: "panel" | "canvas";
+  isCanceled: boolean;
+};
+
+export type DragMovePayload = { canvasCoordinates: Point };
+
+const initialState = {
+  dropTarget: undefined as DropTarget<Instance> | undefined,
+  dragItem: undefined as BaseInstance | undefined,
+};
+
+export const useDragAndDrop = () => {
+  const [rootInstance] = useRootInstance();
+  const [textEditingInstanceId] = useTextEditingInstanceId();
+
+  const state = useRef({ ...initialState });
+
+  const autoScrollHandlers = useAutoScroll({ fullscreen: true });
+
+  const getDefaultDropTarget = () => {
+    const element = rootInstance && getInstanceElementById(rootInstance.id);
+
+    // Should never happen
+    if (!element || !rootInstance) {
+      throw new Error("Could not find root instance element");
+    }
+
+    return { element, data: rootInstance };
+  };
+
+  const dropHandlers = useDrop<Instance>({
+    isDropTarget(element) {
+      const instance =
+        rootInstance !== undefined &&
+        findInstanceByElement(rootInstance, element);
+
+      return instance || false;
+    },
+
+    // This must be fast, it can be called multiple times per pointer move
+    swapDropTarget(dropTarget) {
+      const { dragItem } = state.current;
+
+      if (!dropTarget || dragItem === undefined || rootInstance === undefined) {
+        return getDefaultDropTarget();
+      }
+
+      if (dropTarget.data.id === rootInstance.id) {
+        return dropTarget;
+      }
+
+      const path = getInstancePath(rootInstance, dropTarget.data.id);
+      path.reverse();
+
+      if (dropTarget.nearEdge) {
+        path.shift();
+      }
+
+      // Don't allow to drop inside drag item or any of its children
+      const dragItemIndex = path.findIndex(
+        (instance) => instance.id === dragItem.id
+      );
+      if (dragItemIndex !== -1) {
+        path.splice(0, dragItemIndex + 1);
+      }
+
+      const data = path.find((instance) =>
+        primitives[instance.component].canAcceptChild()
+      );
+
+      if (!data) {
+        return getDefaultDropTarget();
+      }
+
+      if (data.id === dropTarget.data.id) {
+        return dropTarget;
+      }
+
+      const element = data && getInstanceElementById(data.id);
+
+      if (element === null) {
+        return getDefaultDropTarget();
+      }
+
+      return { data, element };
+    },
+
+    onDropTargetChange(dropTarget) {
+      state.current.dropTarget = dropTarget;
+      publish<"dropTargetChange", DropTargetChangePayload>({
+        type: "dropTargetChange",
+        payload: {
+          rect: dropTarget.rect,
+          placement: dropTarget.placement,
+          position: dropTarget.indexWithinChildren,
+          instance: toBaseInstance(dropTarget.data),
+        },
+      });
+    },
+  });
+
+  const dragHandlers = useDrag<Instance>({
+    isDragItem(element) {
```

Comment:
this naming suggests it will return a boolean, but we return an instance in the end